### ISSUE TITLE
Incorrect use-before-declaration error in async IIFE initializer

### ISSUE THREAD

**Issue by Jessidhia** [microsoft/TypeScript#50123]
# Bug Report

### 🔎 Search Terms

- dead zone
- await tdz
- use before declaration await

### 🕗 Version & Regression Information

- This is the behavior in every version I tried, and I reviewed the FAQ for entries about await

### ⏯ Playground Link

[Playground link with relevant code](
https://www.typescriptlang.org/play?ts=4.8.0-dev.20220731#code/MYewdgzgLgBADgJxAWwJYQKYwLwwBQCGEAnmMPgJQ4B8MA3gLABQMMBA7garGAK4A2-Zq0Qp0GZgF8KeCsyA)

### 💻 Code

```ts
const promise = (async () => {
  await null
  promise
})()
```

<details>
<summary>Real life code sample affected by the error</summary>

```tsx
  useEffect(() => {
    const request = (async () => {
      try {
        const response = await premiumApi.announcements(client)
        // @ts-expect-error TypeScript bug, assumes "request" is still in dead zone after await
        if (pendingRequest.current === request || pendingRequest.current === undefined) {
          pendingRequest.current = undefined
          setApiResponse(response)
        }
      } catch {
        pendingRequest.current = undefined
      }
    })()
    pendingRequest.current = request
    return () => {
      pendingRequest.current = undefined
    }
  }, [client])
```

</details>

### 🙁 Actual behavior

- "Block-scoped variable 'promise' used before its declaration."
- "Variable 'promise' is used before being assigned."

### 🙂 Expected behavior

- No errors related to the `promise` variable.

This is not about the "'await' has no effect on the type of this expression" warning, this is a feature and is expected ~~despite me disagreeing with it~~.

The first `await` to be reached in an `async` function is the point where the function stops being synchronous and yields a `Promise` for its result. This means that, immediately after the `await`, the `const promise` is no longer in TDZ and is well-defined: it is the not-yet-resolved `Promise` that was produced by the IIFE's invocation. Now, `await`ing the `promise` _would_ be a deadlock, as the promise would now be waiting for itself to resolve before being able to resolve, but the `Promise` instance itself is still useful for some purposes like `===` checks.

**Comment by RyanCavanaugh** [comment]
I think we can take a simpler fix that is slightly over-pessimistic - the "IIFEs get inlined for CFA" logic shouldn't ever be applied to `async` functions.

**Comment by barak007** [comment]
I have similar case with destructure into `const` bindings with default value,

Playground:
https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAcwKZQCoHc4YBYxjIDOAFAJSIDeAUIvYgE7oiNJWIKoBciAjABpEUHL3AATVMEKpxiAL41FNCAmJRqnMKiEi4iALyIKhgHybVYYnAA2qAHQ24yUl105EAegBUiVAA8AB1RoWWF9HEYAa0RvT0p5BUMUdGxcAiIycgBuGhU1DSgM5ABmZJMDc0trO0dnUiLCUpyvT0RIqOJEaW12ggg8RBgu1GJiUZgAQxsbAE9hPFRhDwAjQnEm4nsgA

**Comment by gabritto** [comment]
> I have similar case with destructure into `const` bindings with default value,
> 
> Playground: https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mABAcwKZQCoHc4YBYxjIDOAFAJSIDeAUIvYgE7oiNJWIKoBciAjABpEUHL3AATVMEKpxiAL41FNCAmJRqnMKiEi4iALyIKhgHybVYYnAA2qAHQ24yUl105EAegBUiVAA8AB1RoWWF9HEYAa0RvT0p5BUMUdGxcAiIycgBuGhU1DSgM5ABmZJMDc0trO0dnUiLCUpyvT0RIqOJEaW12ggg8RBgu1GJiUZgAQxsbAE9hPFRhDwAjQnEm4nsgA

You might want to file a separate issue about that destructuring-specific case (if an issue for that doesn't already exist).

**Comment by barak007** [comment]
@gabritto is #50199 also fixes my issue?

**Comment by gabritto** [comment]
> @gabritto is #50199 also fixes my issue?

I don't think so, the fix was only for async functions and in your case your function is not async. Can you open a new issue for your particular case?

**Comment by artsiommiksiuk** [comment]
Is it really fixed? I'm still reproducing this issue with any version from 4.8.4 up to 5.4.5 (available in playground), for original code snippet and for destructured snippet from @barak007. I'm facing the same issue, and I don't see any resolution.

**Comment by gabritto** [comment]
> Is it really fixed? I'm still reproducing this issue with any version from 4.8.4 up to 5.4.5 (available in playground), for original code snippet and for destructured snippet from @barak007. I'm facing the same issue, and I don't see any resolution.

I think you're right, thanks for pointing that out. The PR seems to have only fixed the problem when `strictNullChecks` is off.
